Kickstarter-iOS/Views/UIView+ShimmerLoading.swift

```diff
@@ -0,0 +1,174 @@
+import Foundation
+import Library
+import ObjectiveC
+import UIKit
+
+private let gradientLayerName = "ksr_shimmer_gradientLayer"
+
+private enum ShimmerConstants {
+  enum Locations {
+    static let start: [NSNumber] = [-1.0, -0.5, 0.0]
+    static let end: [NSNumber] = [1.0, 1.5, 2.0]
+  }
+
+  enum Animation {
+    static let movingAnimationDuration: CFTimeInterval = 1.25
+    static let delayBetweenAnimationLoops: CFTimeInterval = 0.3
+  }
+}
+
+private struct AssociatedKeys {
+  static var isLoading = "isLoading"
+  static var shimmersWhenLoading = "shimmersWhenLoading"
+  static var shimmerLayers = "shimmerLayers"
+}
+
+protocol ShimmerLoading: AnyObject {
+  func startLoading()
+  func stopLoading()
+}
+
+extension UIView {
+  var shimmersWhenLoading: Bool {
+    get {
+      guard let value = objc_getAssociatedObject(self, &AssociatedKeys.shimmersWhenLoading) as? Bool else {
+        return false
+      }
+      return value
+    }
+    set(newValue) {
+      objc_setAssociatedObject(
+        self,
+        &AssociatedKeys.shimmersWhenLoading,
+        newValue,
+        objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN_NONATOMIC
+      )
+    }
+  }
+}
+
+extension ShimmerLoading where Self: UIView {
+  private var isLoading: Bool {
+    get {
+      guard let value = objc_getAssociatedObject(self, &AssociatedKeys.isLoading) as? Bool else {
+        return false
+      }
+      return value
+    }
+    set(newValue) {
+      objc_setAssociatedObject(
+        self, &AssociatedKeys.isLoading, newValue, objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN_NONATOMIC
+      )
+
+      self.updateAnimation()
+    }
+  }
+
+  private var shimmerLayers: [CAGradientLayer] {
+    get {
+      guard
+        let value = objc_getAssociatedObject(self, &AssociatedKeys.shimmerLayers) as? [CAGradientLayer]
+      else { return [] }
+
+      return value
+    }
+    set(newValue) {
+      objc_setAssociatedObject(
+        self,
+        &AssociatedKeys.shimmerLayers,
+        newValue,
+        objc_AssociationPolicy.OBJC_ASSOCIATION_RETAIN_NONATOMIC
+      )
+    }
+  }
+
+  // MARK: - Accessors
+
+  func startLoading() {
+    self.isLoading = true
+  }
+
+  func stopLoading() {
+    self.isLoading = false
+  }
+
+  func layoutGradientLayers() {
+    self.layoutIfNeeded()
+
+    self.shimmerLayers.forEach { layer in
+      layer.frame = layer.superlayer?.bounds ?? .zero
+      layer.setNeedsLayout()
+    }
+  }
+
+  // MARK: - Shimmer Animation
+
+  private func updateAnimation() {
+    guard self.isLoading else {
+      self.shimmerLayers.forEach { $0.removeFromSuperlayer() }
+      return
+    }
+
+    allSubViews(of: self)
+      .filter { $0.shimmersWhenLoading }
+      .forEach { view in
+        let gradientLayer = newGradientLayer(with: view.bounds)
```

Comment:
Would something like `shimmerGradientLayer`, `shimmerAnimation` and `shimmerAnimationGroup` be clearer naming for these funcs?